### PR TITLE
[TASK] Use /index.php as application entry point

### PR DESCRIPTION
**Pull request by bnf**
/typo3/index.php is deprecated and must not longer be used.

See https://forge.typo3.org/issues/87889 and
https://review.typo3.org/c/Packages/TYPO3.CMS/+/74366

Releases: main

### CODE REVIEW

**Comment by sbuerk** [comment]
quick view/first impression:

we need to make this conditional based to ensure working with v12 (state before change) and v13 (with changed entrypoints) - or start breaking the 1 TF version 2 core version support away, which I would try to avoid as this will get a huge cry in the community and extension developers.

**Comment by bnf** [comment]
Hmpf....

Current state is a huge mess for us developers and results in ugly code all over the place (For example all that outdated code in `Resources/Core/Build/Configuration/JSUnit` where our only chance of properly improving the core – adding importmaps – was to ignore that outdated and non-changeable configuration in Testing Framework and craft an own jsunit setup in TYPO3).

Providing *API* compatibility for two versions might be fine, but that doesn't mean one TestingFramework version needs to support multiple TYPO3 versions. Extension authors can simply require-dev `"typo3/testing-framework": "^8 || ^9"`.

Note: I also want to refactor other stuff like the initialization with `SystemEnvironmentBuilder::run(…, SystemEnvironmentBuilder::REQUESTTYPE_CLI)` later on. Reason: That initialization is wrong (tests run with HTTP headers but initialized as CLI) and means we do a bunch of tests based on wrong assumptions (and they fail badly when that wrong assumption is corrected, often because the test data is broken as well. I'd like to get that fixed but I guess that won't be doable with a TF that supports v12 and v13.

So please enlighten me on why we can't simply release TF v9 with the same *API* as v8?
Releasing v9 (or even better name it v13) will also help to simply merge this repo into core as we did with EXT:styleguide and use the same deprecation policy as for our core code.

**Comment by bnf** [comment]
> we need to make this conditional based to ensure working with v12 (state before change) and v13 (with changed entrypoints) 

I've now added such conditions for v12 compat and tested on current 12.4 and main via:

* https://review.typo3.org/c/Packages/TYPO3.CMS/+/82595
* https://review.typo3.org/c/Packages/TYPO3.CMS/+/82596

I still think we should do the split versions with focus on merging TF into core, but I'll discuss that when we've got more time for such things.

**Comment by bmack** [comment]
I'm fine with the change as is. @sbuerk ?

**Comment by lolli42** [comment]
tl;dr: Patch looks good, especially with https://review.typo3.org/c/Packages/TYPO3.CMS/+/82595 being green.

About versioning:
Yes, we can branch-off v8 and start a v9 in main now. But we'll need the patch in v8 as well since v8 needs to continue keeping core v13 compat. That's our current deal: use v8 to run extension tests with both v12 and v13 core. When we branch-off, main can get the b/w compat todo's resolved.

The reason for TF supporting two core versions: This guarantees API compatibility for extensions, which is kinda convenient for extension authors. It also forces TF maintainers to think about compat layers. We basically have a triangle between extension<->TF-api<->core-api here. In case of this patch TF<->core is affected, which forces us to have TF compat at this point to keep extension<->TF clean.

Note I'd love to see a TF core-monorepo merge as well, but this needs a bit of additional infrastructure since monorepo currently can't host casual composer packages. When we merge at some point, we could review the compat problematic along the way.

File: Classes/Core/Acceptance/Helper/Login.php [hunk 1]
```diff
@@ -55,21 +55,21 @@ public function useExistingSession(string $role, float|int $waitTime = 0.5)
 
         $hasSession = $this->_loadSession();
         if ($hasSession && $newUserSessionId !== '' && $newUserSessionId !== $this->getUserSessionId()) {
-            $webDriver->amOnPage('/typo3/index.php');
+            $webDriver->amOnPage('/typo3');
```
Comment (lolli42) [inline Classes/Core/Acceptance/Helper/Login.php:58]: those are ok due to v12 / web server redirects.

File: Classes/Core/Functional/FunctionalTestCase.php [hunk 2]
```diff
@@ -496,15 +497,20 @@ private function createServerRequest(string $url, string $method = 'GET'): Serve
     {
         $requestUrlParts = parse_url($url);
         $docRoot = $this->instancePath;
+
+        // @todo: Remove when dropping support for v12
+        $hasConsolidatedHttpEntryPoint = class_exists(CoreHttpApplication::class);
```
Comment (lolli42) [inline Classes/Core/Functional/FunctionalTestCase.php:502]: that's a simple and straight b/w compat detection strategy. fine with me. and it's smart since v13 core without the entry-point patch does not fail (which would happen when going for a typo3Version check).

File: Classes/Core/Functional/FunctionalTestCase.php [hunk 3]
```diff
@@ -944,19 +950,23 @@ private function retrieveFrontendSubRequestResult(
...
-            Environment::getProjectPath(),
-            Environment::getPublicPath(),
-            Environment::getVarPath(),
-            Environment::getConfigPath(),
-            Environment::getPublicPath() . '/index.php',
-            Environment::isWindows() ? 'WINDOWS' : 'UNIX'
-        );
+        // @todo: Remove when dropping support for v12
+        $hasConsolidatedHttpEntryPoint = class_exists(CoreHttpApplication::class);
+        if (!$hasConsolidatedHttpEntryPoint) {
```
Comment (lolli42) [inline Classes/Core/Functional/FunctionalTestCase.php:955]: So good to see this gone in v13!

**Review by lolli42 — APPROVED** [review]

**Review by sbuerk — APPROVED** [review]

**Comment by bnf** [comment]
thx!
